File: src/main/java/com/stride/DoubleLinkedList.java
package com.stride;

import static java.lang.String.format;

public class DoubleLinkedList<T> {
    private Node<T> head;
    private Node<T> tail;

    private int size;

    public void append(T data) {
        if (isEmpty()) {
            head = new Node<>(null, null, data);
            tail = head;
        } else {
            Node<T> newNode = new Node<>(tail, null, data);
            tail.next(newNode);
            tail = newNode;
        }
        size++;
    }

    public void prepend(T data) {
        if (isEmpty()) {
            head = new Node<>(null, null, data);
            tail = head;
        } else {
            Node<T> newNode = new Node<>(null, head, data);
            head.previous(newNode);
            head = newNode;
        }
        size++;
    }

    public int indexOf(T data) {
        Node pointer = head;
        int result = 0;
        boolean found = false;

        while (pointer.next() != null && !found) {
            if (data == null && pointer.data() == null || data != null && data.equals(pointer.data())) {
                found = true;
            } else {
                pointer = pointer.next();
                result++;
            }
        }
        return result;
    }

    public T remove(int index) {
        Node<T> pointer = nodeAt(index);

        if (pointer == head) {
            head = head.next();
            if (head != null) {
                head.previous(null);
            }
        } else if(pointer == tail) {
            tail = tail.previous();
            tail.next(null);
        } else {
            Node<T> previous = pointer.previous();
            Node<T> next = pointer.next();

            if (previous != null) {
                previous.next(next);
            }
            if (next != null) {
                next.previous(previous);
            }
        }

        size--;
        return pointer.data();
    }

    public T get(int index) {
        return nodeAt(index).data();
    }

    public void set(int index, T data) {
        nodeAt(index).data(data);
    }

    public void clear() {
        head = null;
        tail = null;
        size = 0;
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size() == 0;
    }

    public T first() {
        return head.data();
    }

    public T last() {
        return tail.data();
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder("[");

        Node<T> pointer = head;
        while (pointer != null) {
            builder.append(pointer.toString());
            if (pointer.next() != null) {
                builder.append(", ");
            }
            pointer = pointer.next();
        }
        return builder.append("]").toString();
    }

    private void validateRange(int index) {
        if (isEmpty() || index < 0 || index > size() - 1) {
            throw new IndexOutOfBoundsException(format("Index %d, Size %d", index, size));
        }
    }

    private Node<T> nodeAt(int index) {
        validateRange(index);
        Node<T> pointer;
        if (index < size / 2) {
            pointer = head;
            for (int i = 0; i < index; i++) {
                pointer = pointer.next();
            }
        } else {
            pointer = tail;
            for(int i = size - 1; i > index; i--) {
                pointer = pointer.previous();
            }
        }
        return pointer;
    }
}

File: ruby/lib/double_linked_list.rb
require_relative 'node'

class DoubleLinkedList
  attr_reader :size

  def initialize
    @size = 0
  end

  def append(data)
    if empty?
      @head = Node.new(nil, nil, data)
      @tail = @head
    else
      newNode = Node.new(@tail, nil, data)
      @tail.after = newNode
      @tail = newNode
    end
    @size += 1
  end

  def prepend(data)
    if empty?
      @head = Node.new(nil, nil, data)
      @tail = @head
    else
      newNode = Node.new(nil, @head, data)
      @head.before = newNode
      @head = newNode
    end
    @size += 1
  end

  def index_of(data)
    pointer = @head
    result = 0
    found = false

    while pointer.after && !found
      if data.nil? && pointer.data.nil? || data && data == pointer.data
        found = true
      else
        pointer = pointer.after
        result += 1
      end
    end
    found ? result : -1
  end

  def remove(index)
    pointer = node_at(index)

    if pointer == @head
      @head = @head.after
      @head.before = nil unless @head.nil?
    elsif pointer == @tail
      @tail = @tail.before
      @tail.after = nil
    else
      before = pointer.before
      after = pointer.after
      before.after = after if before
      after.before = before if after
    end

    @size -= 1
    pointer.data
  end

  def [](index)
    node_at(index).data
  end

  def []=(index, data)
    node_at(index).data = data
  end

  def clear
    @head = nil
    @tail = nil
    @size = 0
  end

  def empty?
    size.zero?
  end

  def first
    @head.data
  end

  def last
    @tail.data
  end

  def to_s
    result = '['

    pointer = @head
    while pointer
      result << pointer.to_s
      result << ', ' if pointer.after
      pointer = pointer.after
    end

    result << ']'
  end

  private

  def node_at(index)
    validate_range(index)
    pointer = @head
    (index).times { pointer = pointer.after }
    pointer
  end

  def validate_range(index)
    raise ArgumentError.new("Index ${index}, Size #{@size}") if empty? || index < 0 || index > @size - 1
  end
end